api/workload: Add v2 workload API client

APTy opened the pull request:
**Pull Request check list**

- [ ] Commit conforms to CONTRIBUTING.md?
- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
Adds a workload API that uses a callback-style interface for keeping users' X509SVIDs up to date.

**Description of change**
We've been using this API to interact with SPIFFE Workload API with good results, and thought it was a good time to share with upstream and get feedback. Thanks!

ZymoticB commented on api/workload/v2/stream.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func newConn(ctx context.Context, addr string) (*grpc.ClientConn, error) {
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure())
```
this should also have `grpc.WithBlock()` since its managing all the connection details. Having grpc dialing in the background is confusing.

azdagron commented:
I'm loving the convenience structs that the Client provides with everything all parsed out already. I'm a little concerned about the complexity under the covers. If you are just managing a single connection to the unix socket and a single FetchX509SVID stream, it seems like most of it could be replaced with a for loop to reestablish a connection and open a stream, and an inner for loop to receive and process updates from the stream, with application of backoff in between reconnections.  https://github.com/spiffe/spire/blob/master/api/workload/x509_client.go#L173 has most of what you'd need (it has a little extra goo to support the semantics of the existing Client that wouldn't be necessary).

ZymoticB commented on api/workload/v3/workload.go:
```diff
@@ -0,0 +1,218 @@
+
+// creates single watch loop and returns whether we should continue watching
+func (c *X509SVIDClient) watch() (done bool) {
+	conn, err := c.newConn()
```
I've been spending more time reading the GRPC code and I don't think we need to create a new conn here. GRPC should be managing the subconns in the background (in this case a single connection to a unix socket). Creating a new stream will cause it to hit the balancer which will either succeed because there is an active subconn, or fail because there isn't and we'll need to retry around that.

Additionally, given we need to implement that retry logic, we *may* not want grpc.WithBlock. I'll need to go re-read the docs there. We may still want it to enable the client's context to manage the conn.

APTy replied:
removed WithBlock so that we can implement the initial connection retry ourselves and surface the error to users

ZymoticB commented on api/workload/v3/workload.go:
```diff
@@ -0,0 +1,218 @@
+
+func (c *X509SVIDClient) handleX509SVIDStream(stream workload.SpiffeWorkloadAPI_FetchX509SVIDClient) error {
+	for {
+		resp, err := stream.Recv()
```
Am I correct in saying that for this to terminate we depend on grpc.WithBlock to enable the context to control the conn which will cancel the stream?

I guess the grpcCtx above is coming from that c.ctx so maybe we are okay either way.

APTy replied:
yeah this gets canceled because of the grpcCtx that chains up to the main context

APTy commented on api/workload/v3/backoff.go:
```diff
@@ -0,0 +1,34 @@
+package workload
```
v3 is unintentional

APTy commented:
I've simplified the implementation significantly and rebased over master to account for some of the package changes. One of my final questions is how we want to allow this interface to grow in the future.

Currently, I've named the primary watcher interface `WorkloadX509SVIDWatcher`, since it is opinionated to X509 SVIDs, and cannot change to also include JWTs (due to backwards-compatibility of interfaces). If we want to add JWTs later, it would need to be a new client altogether.

If we want to maintain a single client for many SVID types, should we expose a more generic `WorkloadSVIDWatcher` that has a `UpdateSVID(*SVID)` method? (Where `*SVID` is a struct that includes a `X509` or a `JWT` type)

azdagron commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,219 @@
+	}
+}
+
+// creates single watch loop and returns whether we should continue watching
```
stale comment about the return

azdagron commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,219 @@
+		}
+		svids, err := protoToX509SVIDs(resp)
+		if err != nil {
+			return err
```
returning an error here will cause backoff on an otherwise healthy stream. should this just call OnError() and `continue`?

More importantly, if the stream does not Recv() everything, it will stay active and we'll essentially leak the stream until the context passed into `FetchX509SVID` is canceled. I'd recommend `watch` create a cancellable context from c.ctx that it passes into newX509SvidStream that it can cancel when `handleX509SVIDStream` returns just to make sure we don't leak streams.

APTy replied:
ah, good calll

APTy replied:
i need to add some tests for this

azdagron commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,219 @@
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
```
Start() cannot be called more than once (results in two or more streams) and cannot be called after Stop() (since `run` will return immediately as the context will is cancelled). We should at a minimum document this.

APTy replied:
i've added docs and some state management around this

amoore877 commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,219 @@
+	SPIFFEID     string
+	PrivateKey   crypto.Signer
+	Certificates []*x509.Certificate
+	TrustBundle  *x509.CertPool
```
CertPool is convenient for verify opts, but highly inconvenient for accessing the certs themselves (unexported field). Better to have the trust bundle field be a cert array, though could have convenience method that gives you a cert pool if desired.

APTy replied:
I've changed `TrustBundle` to be a list of certs, with a `TrustBundePool() *x509.CertPool` convenience method

APTy commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,219 @@
+// establishes a new persistent connection, returns nil if a client can't be created within the context deadline
+func (c *X509SVIDClient) newConn() *grpc.ClientConn {
+	for {
+		conn, err := grpc.DialContext(c.ctx, c.addr, grpc.WithInsecure())
```
that's good to know. and while it may be true today, it's not necessarily a contract of the DialContext API, so we still want to check for the error, right?

azdagron replied:
We still want to check the error for sure. However, without WithBlock, this means that the "OnConnected(true)" will be misleading since the code won't now if the async dial has even succeeded by then. You don't really know that you've connected until FetchX509SVID has returned a stream.

ZymoticB commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,239 @@
+	c.watch(conn)
+}
+
+// establishes a new persistent connection, returns nil if a client can't be created within the context deadline
```
this comment is out of date, there is no longer a connection timeout.

ZymoticB commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,239 @@
+	PrivateKey      crypto.Signer
+	Certificates    []*x509.Certificate
+	TrustBundle     []*x509.Certificate
+	trustBundlePool *x509.CertPool
```
Why isn't this exported? The accessor method doesn't seem useful.

azdagron commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,235 @@
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		addr:         DefaultAgentAddress,
```
If the address option isn't set, we should try to pull the address from SPIFFE_ENDPOINT_SOCKET envvar.

From the [Workload API Endpoint Spec](https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Workload_Endpoint.md#4-locating-the-endpoint)
> Clients may be explicitly configured with the socket location, or may utilize the well-known environment variable SPIFFE_ENDPOINT_SOCKET. If not explicitly configured, conforming clients MUST fall back to the environment variable.

APTy replied:
nice, thanks for catching that

azdagron commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,235 @@
+			}
+			continue
+		}
+		if err := c.handleX509SVIDStream(stream); err != nil {
```
right now handleX509SVIDStream always returns an error. However, if a future refactor changes that, the context you created above will not be cancelled and we'll leak contexts. Maybe something like:

```
func watch(conn) {
    for {
        if done := watchOnce(conn); done {
            return true
        }
    }
}

func watchOnce(conn) {
    ctx, cancel := ...
    defer cancel()

    // rest of the body
}
```

azdagron commented on api/workload/v2/state.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+type clientStateManager struct {
+	sync.RWMutex
```
nit:  sync.Mutex since we never take a read lock?

azdagron commented on api/workload/v2/workload.go:
```diff
@@ -0,0 +1,244 @@
+	// FederatedBundles are CA certificate bundles belonging to foreign Trust Domains
+	// that the workload should trust, keyed by the SPIFFE ID of the foreign domain.
+	// Unimplemented.
+	FederatedBundles map[string]*x509.CertPool
```
hmm, should this also be []*x509.Certificate instead?

APTy replied:
hmm yes i think so

amoore877 commented on api/workload/v2/backoff.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// backoff defines an linear backoff policy.
+type backoff struct {
```
Would it make sense to move this to some common location? This seems like it is very generalized and may be useful elsewhere.

azdagron replied:
That's a good idea. I think we can submit this as is and do that refactor once we have the need for it.
